### ISSUE TITLE
Implement AVOID annotating with Function.

### ISSUE THREAD

**Issue by JPaulsen** [dart-lang/linter#554]
From [Effective Dart](https://www.dartlang.org/guides/language/effective-dart/design#avoid-annotating-with-function)

**Comment by JPaulsen** [comment]
@bwilkerson @alexeieleusis I think this rule can't be implemented, I was thinking about the most basic of the cases, a Function parameter with only one use of Functions.apply inside and no other reference to the function but even in that case there would be false positives.

```dart
void applyWithNullArguments(Function f, int n) {
  final arguments = [];
  while (n --> 0) {
    arguments.add(null);
  }
  Function.apply(f, arguments);
}
```

So I think we should discard this one, what do you think?

**Comment by alexeieleusis** [comment]
Probably not linting those cases?

I would like to know if @munificent has some advice here.

**Comment by JPaulsen** [comment]
@alexeieleusis but I think this is one of the most basic cases, when you have only one call to apply, only case it would be a Function with 0 calls to apply but I do not know if that would have any sense in the code.

Other case could be limit the rule to some variable assignations and declarations, for example a final variable declaration like this:

```dart
final Function a = something;
```

So if we can find the type of ```something``` statically and it is better than Function then we could lint that.
